File: 백준/Silver/15650. N과 M （2）/N과 M （2）.py
n,m=map(int,input().split())

visited=[0]*(n+1)
arr=[]

def dfs(start):
    if len(arr) == m:
        print(*arr)
        return
    for i in range (start,n+1):
        if visited[i]:
            continue
        visited[i]=True
        arr.append(i)
        dfs(i+1)
        visited[i]=False
        arr.pop()
dfs(1)

File: 백준/Silver/15650. N과 M （2）/N과 M （2）.java
import java.io.*;
import java.util.*;

public class Main {
	static int n;
	static int m;
	static List<Integer> arr = new LinkedList<>();
	static int start;
	
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		dfs(1);
	
	}
	static void dfs(int start) {
		if (arr.size() == m) {
			for (int num : arr) {
				System.out.print(num+ " ");
			}
			System.out.println();
			return;
		}
		for (int i =start; i < n+1; i++) {
			if (!arr.contains(i)) {
				arr.add(i);
				dfs(i+1);
				arr.remove((Integer)i);
			}
			
		}
	}
}
